vertexai/internal/sentencepiece/encoder.go

```diff
@@ -0,0 +1,329 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sentencepiece
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"unicode/utf8"
+
+	"cloud.google.com/go/vertexai/internal/sentencepiece/internal/model"
+	"cloud.google.com/go/vertexai/internal/sentencepiece/internal/prefixmatcher"
+	"cloud.google.com/go/vertexai/internal/sentencepiece/internal/priorityqueue"
+	"google.golang.org/protobuf/proto"
+)
+
+const debugEncode = false
+
+// Encoder represents a SentencePiece encoder (tokenizer).
+// An Encoder converts input text into a sequence of tokens LLMs use.
+type Encoder struct {
+	model *model.ModelProto
+
+	pieces   map[string]int
+	reserved map[string]int
+
+	// unknownID is the token identifier of the UNKNOWN piece
+	unknownID int
+
+	// userDefinedMatcher is a prefix matcher for symbols that are of
+	// "user-defined" type in the model proto.
+	userDefinedMatcher *prefixmatcher.PrefixMatcher
+
+	// byteTokens is a cache of byte values and the tokens they represent
+	byteTokens map[byte]Token
+}
+
+// NewEncoderFromPath creates a new Encoder from a file path to the protobuf
+// data.
+func NewEncoderFromPath(protoFile string) (*Encoder, error) {
+	f, err := os.Open(protoFile)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %q: %v", protoFile, err)
+	}
+	defer f.Close()
+	return NewEncoder(f)
+}
+
+// NewEncoder creates a new Encoder from a reader with the protobuf data.
+func NewEncoder(protoReader io.Reader) (*Encoder, error) {
+	b, err := io.ReadAll(protoReader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read protobuf data: %v", err)
+	}
+
+	var mp model.ModelProto
+	err = proto.Unmarshal(b, &mp)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal protobuf: %v", err)
+	}
+
+	tspec := mp.GetTrainerSpec()
+	if tspec.GetModelType() != model.TrainerSpec_BPE {
+		return nil, fmt.Errorf("model type %s not supported", tspec.GetModelType())
+	}
+
+	userDefined := make(map[string]struct{})
```

Comment:
Yep, I know. `map[string]struct{}` is the "idiomatic" Go set though, and what Ian uses in his examples and posts (if we ever get a `sets` package, that's what it will have). The extra verbosity here isn't that much of a big deal after all

Reply:
Disagree about the idiomatic but OK.

Reply:
Done